pkg/controller/jobs/kubeflow/trainjob/trainjob_controller.go

```diff
@@ -0,0 +1,250 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trainjob
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kftrainerv2api "github.com/kubeflow/trainer/v2/pkg/apis/trainer/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	jobsetapi "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	kJobset "sigs.k8s.io/kueue/pkg/controller/jobs/jobset"
+	"sigs.k8s.io/kueue/pkg/features"
+	"sigs.k8s.io/kueue/pkg/podset"
+)
+
+var (
+	gvk                    = kftrainerv2api.GroupVersion.WithKind("TrainJob")
+	FrameworkName          = "trainer.kubeflow.org/trainjob"
+	TrainJobControllerName = "trainer.kubeflow.org/trainjob-controller"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:      SetupIndexes,
+		NewJob:            NewJob,
+		NewReconciler:     NewReconciler,
+		SetupWebhook:      SetupTrainJobWebhook,
+		JobType:           &kftrainerv2api.TrainJob{},
+		AddToScheme:       kftrainerv2api.AddToScheme,
+		MultiKueueAdapter: &multiKueueAdapter{},
+	}))
+}
+
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs/finalizers,verbs=get;update
+
+type trainJobReconciler struct {
+	jr     *jobframework.JobReconciler
+	client client.Client
+}
+
+var reconciler trainJobReconciler
+var _ jobframework.JobReconcilerInterface = (*trainJobReconciler)(nil)
+
+func NewReconciler(client client.Client, eventRecorder record.EventRecorder, opts ...jobframework.Option) jobframework.JobReconcilerInterface {
+	reconciler = trainJobReconciler{
+		jr:     jobframework.NewReconciler(client, eventRecorder, opts...),
+		client: client,
+	}
+	return &reconciler
+}
+
+func (r *trainJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	jobSet := jobsetapi.JobSet{}
+	err := r.client.Get(ctx, req.NamespacedName, &jobSet)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
```

Comment:
Looking at the appwrapper code, we just ignore not found entirely.

I worry that this could cause an infinite loop where JobSet is not found we get reconciling and we never bail out.

Reply:
yeah, you're right.  I've changed the implementation to make the controller `Own` the `JobSet` and ignore not found